File: Leetcode/sbgo/Leetcode28-实现strstr.go
package sbgo

func strStr(haystack string, needle string) int {
	n, m := len(haystack), len(needle)
	if m == 0 {
		return 0
	}
	var i, j int
	for i, j = 0, 0; i < n && j < m; i++ {
		if haystack[i] == needle[j] {
			j++
		} else {
			i -= j
			j = 0
		}
	}
	if j < m {
		return -1
	}
	return i - m
}

File: Leetcode/PY/Leetcode28-实现strstr.py
class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        m,n = len(haystack),len(needle)
        if n == 0:
            return 0
        i,j = 0,0
        while i < m and j < n:
            if haystack[i] == needle[j]:
                j += 1
            else:
                i -= j
                j = 0
            i += 1
        if j < n:
            return -1
        return i - n